File: _python-archive/devctl/commands/env.py
"""`devctl env` — scaffold and register projects in the local index.

Heavy imports (rich) are deferred into the command bodies so importing this
module — which cli.py must do to register the sub-app — costs only ``import
typer``.  Commands that never render a table (``add``, ``rm``) skip the ~45 ms
``rich.table`` import entirely.
"""
from __future__ import annotations

from pathlib import Path

import typer

app = typer.Typer(help="Manage your project registry.")


# Minimal language-specific scaffolds — kept small on purpose.
SCAFFOLDS: dict[str, dict[str, str]] = {
    "python": {
        "README.md": "# {name}\n",
        "pyproject.toml": '[project]\nname = "{name}"\nversion = "0.1.0"\n',
        "src/{name}/__init__.py": "",
        ".gitignore": "__pycache__/\n.venv/\n*.pyc\n",
    },
    "node": {
        "README.md": "# {name}\n",
        "package.json": '{{\n  "name": "{name}",\n  "version": "0.1.0"\n}}\n',
        ".gitignore": "node_modules/\n.env\n",
    },
    "generic": {"README.md": "# {name}\n", ".gitignore": ".env\n"},
}

# Files that, if present in a directory, reveal its language — used by `add`
# to auto-tag projects when the user doesn't pass --lang.
_LANG_MARKERS: list[tuple[str, str]] = [
    ("pyproject.toml", "python"),
    ("setup.py", "python"),
    ("requirements.txt", "python"),
    ("package.json", "node"),
    ("Cargo.toml", "rust"),
    ("go.mod", "go"),
]


def _detect_lang(path: Path) -> str | None:
    """Best-effort language detection from well-known marker files."""
    for marker, lang in _LANG_MARKERS:
        if (path / marker).exists():
            return lang
    return None


@app.command("new")
def new(
    name: str,
    lang: str = typer.Option("generic", help="python | node | generic"),
    root: Path = typer.Option(
        Path.cwd(), help="Parent directory to create the project in."
    ),
) -> None:
    """Scaffold a new project folder and auto-register it."""
    from rich.console import Console

    from .. import db

    console = Console()
    if lang not in SCAFFOLDS:
        console.print(f"[red]Unknown lang '{lang}'.[/] Choose: {', '.join(SCAFFOLDS)}")
        raise typer.Exit(1)

    project_path = (root / name).resolve()
    if project_path.exists():
        console.print(f"[red]{project_path} already exists.[/]")
        raise typer.Exit(1)

    for rel, template in SCAFFOLDS[lang].items():
        file_path = project_path / rel.format(name=name)
        file_path.parent.mkdir(parents=True, exist_ok=True)
        file_path.write_text(template.format(name=name))

    db.add_project(name, project_path, lang)
    console.print(f"[green]✓[/] Created and registered [bold]{name}[/] at {project_path}")


@app.command("add")
def add(
    path: Path,
    name: str = typer.Option(None, help="Defaults to folder name."),
    lang: str = typer.Option(None, help="Optional language tag (auto-detected if omitted)."),
) -> None:
    """Register an existing folder as a project (language auto-detected)."""
    from rich.console import Console

    from .. import db

    console = Console()
    path = path.resolve()
    if not path.is_dir():
        console.print(f"[red]{path} is not a directory.[/]")
        raise typer.Exit(1)
    resolved_lang = lang or _detect_lang(path)
    db.add_project(name or path.name, path, resolved_lang)
    tag = f" [magenta]({resolved_lang})[/]" if resolved_lang else ""
    console.print(f"[green]✓[/] Registered [bold]{name or path.name}[/]{tag}")


@app.command("rm")
def rm(name: str) -> None:
    """Remove a project from the registry (does NOT delete files)."""
    from rich.console import Console

    from .. import db

    console = Console()
    if db.remove_project(name):
        console.print(f"[green]✓[/] Removed [bold]{name}[/]")
    else:
        console.print(f"[yellow]No project named '{name}'.[/]")


@app.command("ls")
def ls() -> None:
    """List all registered projects, ranked by frecency."""
    from rich.console import Console
    from rich.table import Table

    from .. import db

    console = Console()
    rows = db.list_projects()
    if not rows:
        console.print("[dim]No projects yet.[/]")
        return
    table = Table(title="Projects", show_lines=False)
    table.add_column("Name", style="cyan")
    table.add_column("Lang", style="magenta")
    table.add_column("Uses", style="green", justify="right")
    table.add_column("Path", style="white")
    for r in rows:
        table.add_row(r["name"], r["lang"] or "-", str(r["use_count"]), r["path"])
    console.print(table)

File: src/commands/env.rs
//! `devctl env` — scaffold and register projects in the local index.

use std::path::PathBuf;

use anyhow::Result;
use comfy_table::{presets::UTF8_FULL, Cell, Table};

use crate::db;
use crate::ui;
use crate::util;

/// Minimal language-specific scaffolds — kept small on purpose. Each entry is
/// `(relative_path, template)`; `{name}` is substituted in both.
fn scaffolds(lang: &str) -> Option<Vec<(&'static str, &'static str)>> {
    match lang {
        "python" => Some(vec![
            ("README.md", "# {name}\n"),
            (
                "pyproject.toml",
                "[project]\nname = \"{name}\"\nversion = \"0.1.0\"\n",
            ),
            ("src/{name}/__init__.py", ""),
            (".gitignore", "__pycache__/\n.venv/\n*.pyc\n"),
        ]),
        "node" => Some(vec![
            ("README.md", "# {name}\n"),
            (
                "package.json",
                "{\n  \"name\": \"{name}\",\n  \"version\": \"0.1.0\"\n}\n",
            ),
            (".gitignore", "node_modules/\n.env\n"),
        ]),
        "generic" => Some(vec![("README.md", "# {name}\n"), (".gitignore", ".env\n")]),
        _ => None,
    }
}

/// Files that, if present in a directory, reveal its language.
const LANG_MARKERS: &[(&str, &str)] = &[
    ("pyproject.toml", "python"),
    ("setup.py", "python"),
    ("requirements.txt", "python"),
    ("package.json", "node"),
    ("Cargo.toml", "rust"),
    ("go.mod", "go"),
];

fn detect_lang(path: &std::path::Path) -> Option<String> {
    for (marker, lang) in LANG_MARKERS {
        if path.join(marker).exists() {
            return Some((*lang).to_string());
        }
    }
    None
}

pub fn new(name: &str, lang: &str, root: Option<PathBuf>) -> Result<i32> {
    let Some(files) = scaffolds(lang) else {
        println!(
            "{} Choose: python, node, generic",
            ui::red(&format!("Unknown lang '{lang}'."))
        );
        return Ok(1);
    };

    let root = match root {
        Some(r) => r,
        None => std::env::current_dir()?,
    };
    let project_path = root.join(name);
    if project_path.exists() {
        println!(
            "{}",
            ui::red(&format!("{} already exists.", project_path.display()))
        );
        return Ok(1);
    }

    for (rel, template) in files {
        let rel = rel.replace("{name}", name);
        let file_path = project_path.join(&rel);
        if let Some(parent) = file_path.parent() {
            std::fs::create_dir_all(parent)?;
        }
        std::fs::write(&file_path, template.replace("{name}", name))?;
    }

    let project_path = util::resolve_path(&project_path);
    db::add_project(
        name,
        &project_path.to_string_lossy(),
        Some(lang.to_string()),
    )?;
    println!(
        "{} Created and registered {} at {}",
        ui::green("✓"),
        ui::bold(name),
        project_path.display()
    );
    Ok(0)
}

pub fn add(path: PathBuf, name: Option<String>, lang: Option<String>) -> Result<i32> {
    let path = util::resolve_path(&path);
    if !path.is_dir() {
        println!(
            "{}",
            ui::red(&format!("{} is not a directory.", path.display()))
        );
        return Ok(1);
    }
    let resolved_lang = lang.or_else(|| detect_lang(&path));
    let proj_name = name.unwrap_or_else(|| {
        path.file_name()
            .map(|n| n.to_string_lossy().to_string())
            .unwrap_or_default()
    });
    db::add_project(&proj_name, &path.to_string_lossy(), resolved_lang.clone())?;
    let tag = match &resolved_lang {
        Some(l) => format!(" {}", ui::magenta(&format!("({l})"))),
        None => String::new(),
    };
    println!(
        "{} Registered {}{}",
        ui::green("✓"),
        ui::bold(&proj_name),
        tag
    );
    Ok(0)
}

pub fn rm(name: &str) -> Result<i32> {
    if db::remove_project(name)? {
        println!("{} Removed {}", ui::green("✓"), ui::bold(name));
    } else {
        println!("{}", ui::yellow(&format!("No project named '{name}'.")));
    }
    Ok(0)
}

pub fn ls() -> Result<i32> {
    let rows = db::list_projects()?;
    if rows.is_empty() {
        println!("{}", ui::dim("No projects yet."));
        return Ok(0);
    }
    let mut table = Table::new();
    table.load_preset(UTF8_FULL);
    table.set_header(vec!["Name", "Lang", "Uses", "Path"]);
    for r in &rows {
        table.add_row(vec![
            Cell::new(&r.name),
            Cell::new(r.lang.as_deref().unwrap_or("-")),
            Cell::new(r.use_count),
            Cell::new(&r.path),
        ]);
    }
    println!("{table}");
    Ok(0)
}
